http/encoding.go

```diff
@@ -97,22 +103,58 @@ func ResponseEncoder(ctx context.Context, w http.ResponseWriter) Encoder {
 		}
 		return nil, ""
 	}
+	negotiateCT := func(ct string) Encoder {
+		switch {
+		case ct == "application/json" || strings.HasSuffix(ct, "+json"):
+			return json.NewEncoder(w)
+		case ct == "application/xml" || strings.HasSuffix(ct, "+xml"):
+			return xml.NewEncoder(w)
+		case ct == "application/gob" || strings.HasSuffix(ct, "+gob"):
+			return gob.NewEncoder(w)
+		default:
+			return json.NewEncoder(w)
+		}
+	}
 	var accept string
-	if a := ctx.Value(AcceptTypeKey); a != nil {
-		accept = a.(string)
+	{
+		if a := ctx.Value(AcceptTypeKey); a != nil {
+			accept = a.(string)
+		}
+	}
+	var ct string
+	{
+		if a := ctx.Value(ContentTypeKey); a != nil {
+			ct = a.(string)
+		}
 	}
 	var (
 		enc Encoder
 		mt  string
+		err error
 	)
-	if enc, mt = negotiate(accept); enc == nil {
-		// attempt to normalize
-		if mt, _, err := mime.ParseMediaType(accept); err == nil {
-			enc, mt = negotiate(mt)
+	{
+		// If Accept header exists in the request, infer the response encoder
+		// from the header value.
+		if enc, mt = negotiate(accept); enc == nil {
+			// attempt to normalize
+			if mt, _, err = mime.ParseMediaType(accept); err == nil {
+				enc, mt = negotiate(mt)
+			}
+			if enc != nil {
+				SetContentType(w, mt)
+				return enc
+			}
+		}
+		if ct != "" {
```

Comment:
This should probably take precedence over the content type negotiation using the Accept header.